### PR TITLE
feat: add option to skip write when possible

### LINKED ISSUE #3: feat: track session changes via deep proxification
> I suggest slightly reworking the session persistence condition. Currently, on any read, we write to persist the session, which produces unnecessary write operations. We can deeply proxy the session object to track mutations more precisely.
> 
> This is just a raw example of proxying; it can be adjusted only for objects and arrays (JSON serializable) and with `if`'s for performance reasons.
> ```ts
> const mutated = Symbol("mutated");
> 
> // for Set, Map, and Array
> const mutation_methods = new Set([
>   "set",
>   "add",
>   "delete",
>   "clear",
>   "push",
>   "pop",
>   "shift",
>   "unshift",
>   "splice",
>   "sort",
>   "reverse",
>   "fill",
>   "copyWithin",
> ]);
> 
> /**
>  * Creates a proxy wrapper around an object to track mutations and handle cleanup
>  * @param {any} object - The target object to proxify
>  * @param {Function} [dispose] - Optional disposal function that runs during cleanup
>  * @returns {Proxy} Proxied object that tracks mutations
>  *
>  * @example
>  * const obj = { count: 0 };
>  * const proxiedObj = proxify(obj, () => {
>  *   console.log('cleanup');
>  * });
>  */
> export function proxify<T>(
>   object: any,
>   dispose?: () => Promise<void> | void,
> ): T & { [Symbol.asyncDispose]: () => Promise<void>; [mutated]: boolean } {
>   object[Symbol.asyncDispose] = dispose;
>   object[mutated] = false;
> 
>   const handler: ProxyHandler<any> = {
>     get(target, key: string) {
>       if (typeof target[key] === "object" && target[key] !== null) {
>         return new Proxy(target[key], handler);
>       }
>       if (typeof target[key] === "function") {
>         return (...args: unknown[]) => {
>           if (mutation_methods.has(key)) {
>             object[mutated] = true;
>           }
>           return target[key](...args);
>         };
>       }
>       return target[key];
>     },
>     set(target, prop: string, value) {
>       object[mutated] = true;
>       target[prop] = value;
>       return true;
>     },
>   };
> 
>   return new Proxy(object, handler);
> }
> 
> proxify.is_mutated = mutated;
> ```
> 
> usage with `using`: 
> ```ts
> async function app() {
>   await using tracked_state = proxify(state, async () => {
>     if (state[proxify.is_mutated]) {
>       console.log("saving state");
>       state[proxify.is_mutated] = false;
>       await db.set(["state"], state);
>     }
>   });
> 
> 
>   tracked_state.a = 1; // -> mutation
> 
> } // -> save
> ``` 

 _Originally posted by @shevernitskiy in [#712](https://github.com/grammyjs/grammY/issues/712#issuecomment-2629472714)_

### PR DESCRIPTION
**Pull request by roziscoding**
I need help in figuring out other possible edge cases or caveats so we can add tests for them and, ultimately, document them.

Usage:

```ts
composer.use(session({ trackChanges: true })
```

Closes #3

### CODE REVIEW

File: proxy.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
+/**
+ * @internal Symbol used to track if the object has been mutated
+ */
+export const _MUTATED = Symbol("mutated");
+
+/**
+ * @internal Symbol used to track if the object is being tracked
+ */
+export const _TRACKED = Symbol("tracked");
```
Comment (KnorpelSenf) [inline proxy.ts:9]: Can we leave them internal, or is there a good reason to export them?
```suggestion
/**
 * @internal Symbol used to track if the object has been mutated
 */
const _MUTATED = Symbol("mutated");

/**
 * @internal Symbol used to track if the object is being tracked
 */
const _TRACKED = Symbol("tracked");
```

File: proxy.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+ * @internal Creates a proxy wrapper around an object to track mutations
+ * @param object - The target object to proxify
+ * @returns Proxied object that tracks mutations
+ */
+export function trackMutations<T>(
+    object: T,
+): Mutatable<T> {
+    if (typeof object !== "object" || object === null) {
+        return object as Mutatable<T>;
+    }
```
Comment (KnorpelSenf) [inline proxy.ts:40]: So … strings are returned as-is? Is this correct?

File: proxy.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+    const proxyCache = new WeakMap();
```
Comment (KnorpelSenf) [inline proxy.ts:42]: How hard would it be to tie the cache to a property session instead? That would feel like a more obvious strategy to manage the lifetime of these objects in memory

File: proxy.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    let hasChanges = false;
+
+    // deno-lint-ignore no-explicit-any
+    const handler: ProxyHandler<any> = {
+        get(target, prop) {
+            if (prop === _TRACKED) return true;
+            if (prop === _MUTATED) return hasChanges;
+            const originalValue = Reflect.get(target, prop);
+            if (typeof originalValue === "object" && originalValue !== null) {
+                if (proxyCache.has(originalValue)) {
```
Comment (KnorpelSenf) [inline proxy.ts:53]: Can we skip further proxification if `hasChanges` is already `true`? It might be a bit of a micro-optimisation so I would understand if you dislike this

**Review by KnorpelSenf — CHANGES_REQUESTED** [review]

File: proxy.ts (hunk 2 above)
Comment (roziscoding) [reply proxy.ts:40]: It is. There's nothing to track besides direct assignment and that's done on the parent object. Also:
CleanShot 2025-04-06 at 09 35 20@2x

File: proxy.ts (hunk 3 above)
Comment (roziscoding) [reply proxy.ts:42]: I don't think it'd be that hard, but it's not obvious to me why we'd want to do it, sorry.

File: proxy.ts (hunk 4 above)
Comment (roziscoding) [reply proxy.ts:59]: Good idea!

File: proxy.ts (hunk 1 above)
Comment (roziscoding) [reply proxy.ts:9]: We use them for testing.
I don't know if there's another way of exposing these to tests but not to the public API.

Comment (KnorpelSenf) [reply proxy.ts:9]: I think we can use `Symbol.for` then

Comment (KnorpelSenf) [reply proxy.ts:9]: Or just make sure we only export things internally but don't leak it to package consumers :)

File: proxy.ts (hunk 3 above)
Comment (KnorpelSenf) [reply proxy.ts:42]: I find it hard to reason about the lifecycle of memory when there are different references to the objects, and only some of them prevent it from being garbage-collected. This is especially so if there is an obvious way to structure that code such that the memory is directly tied to the stack, which makes it very obvious how long the objects live. But if you have strong opinions about keeping it in a `WeakMap`, then I'll be able to wrap my head around it :)

File: proxy.ts (hunk 4 above)
Comment (KnorpelSenf) [reply proxy.ts:59]: Fixed in 21cf040, resolving
